File: 3-oauth2-google-login/backend/0-typescript/src/modules/auth/auth.service.ts
/**
 * Business logic service for Auth.
 * Owns the two responsibilities that sit between the Google OAuth flow and the database:
 * 1. Federated identity upsert — map a Google profile to a local user row.
 * 2. JWT issuance — sign an internal access token after the identity is confirmed.
 */
import {
    Injectable,
} from "@nestjs/common"
import {
    JwtService,
} from "@nestjs/jwt"
import {
    InjectRepository,
} from "@nestjs/typeorm"
import {
    Repository,
} from "typeorm"
import {
    UserEntity,
} from "../../entities"
import type {
    GoogleProfilePayload,
} from "./google-profile"

/**
 * Return type for {@link AuthService.findOrCreateGoogleUser}.
 * Carries both the persisted entity and a flag indicating whether a new row was inserted.
 */
export interface FindOrCreateResult {
    /** The persisted (or updated) user entity. */
    user: UserEntity
    /**
     * `true` when a new row was INSERTed (first-time login);
     * `false` when an existing row was found and updated.
     */
    isNewUser: boolean
}

/**
 * Links Google profiles to local DB rows and issues internal JWTs.
 * Stateless service — no session or request-scoped state is held here.
 */
@Injectable()
export class AuthService {
    constructor(
    @InjectRepository(UserEntity)
    private readonly usersRepo: Repository<UserEntity>,
    private readonly jwtService: JwtService,
    ) {}

    /**
     * Upserts a local user row linked to the provided Google identity.
     *
     * Lookup strategy (two-stage):
     * 1. Search by `googleId` — the stable provider identifier for returning users.
     * 2. If not found, search by `email` — to silently link an existing password account with Google.
     * 3. If neither exists, INSERT a new row with `password = null` (OAuth-only account).
     *
     * @param payload — normalized Google profile from the Passport strategy or mock branch.
     * @returns persisted user entity + `isNewUser` flag for onboarding routing on the client.
     */
    async findOrCreateGoogleUser(payload: GoogleProfilePayload): Promise<FindOrCreateResult> {
        // Stage 1: try stable googleId first — fastest path for returning users.
        // Stage 2 (via ??): fall back to email linking so a password account can gain Google login.
        let user =
      (await this.usersRepo.findOne({
          where: {
              googleId: payload.googleId,
          },
      })) ??
      (await this.usersRepo.findOne({
          where: {
              email: payload.email,
          },
      }))

        if (!user) {
            // First time we see this identity — INSERT a new OAuth-only user row.
            // password is null because the credential lives at Google, not here.
            user = this.usersRepo.create({
                email: payload.email,
                googleId: payload.googleId,
                firstName: payload.firstName ?? null,
                lastName: payload.lastName ?? null,
                picture: payload.picture ?? null,
                password: null,
            })
            await this.usersRepo.save(user)
            return {
                user,
                isNewUser: true,
            }
        }

        // Returning user — enrich any updated profile fields from Google.
        // We always re-link the googleId (handles the email-matched case where googleId was null).
        user.googleId = payload.googleId
        // Only overwrite name/picture if Google provided a value — preserve existing data otherwise.
        user.firstName = payload.firstName ?? user.firstName
        user.lastName = payload.lastName ?? user.lastName
        user.picture = payload.picture ?? user.picture
        await this.usersRepo.save(user)
        return {
            user,
            isNewUser: false,
        }
    }

    /**
     * Convenience wrapper — returns just the user entity (drops the `isNewUser` flag).
     * Kept for backward compatibility with call sites that don't need the flag.
     *
     * @param payload — normalized Google profile payload.
     * @returns persisted user entity.
     */
    async findOrCreateFromGoogle(payload: GoogleProfilePayload): Promise<UserEntity> {
        const { user } = await this.findOrCreateGoogleUser(payload)
        return user
    }

    /**
     * Signs an internal JWT and formats the final JSON response returned to the client.
     * The JWT carries only `sub` (user.id) — all other claims are fetched fresh from DB
     * on each authenticated request, so there is no stale-claim problem.
     *
     * @param user — persisted user entity (used for JWT subject and response body).
     * @param isNewUser — propagated from {@link findOrCreateGoogleUser}; lets the client branch on first login.
     * @returns object with `message`, `access_token`, `isNewUser`, and a safe `user` summary.
     */
    async completeGoogleLogin(user: UserEntity, isNewUser: boolean): Promise<{
        message: string
        access_token: string
        isNewUser: boolean
        user: { id: number; email: string; name: string; googleSub: string | null }
    }> {
        // Sign JWT with only the user id as sub — minimal payload reduces token size.
        const access_token = await this.jwtService.signAsync({
            sub: user.id,
        })
        return {
            message: "Đăng nhập Google thành công!",
            access_token,
            isNewUser,
            user: {
                id: user.id,
                email: user.email,
                // Concatenate first + last, filtering null/undefined to avoid "null null".
                name: [user.firstName,
                    user.lastName,
                ].filter(Boolean).join(" "),
                // Expose googleId under the alias `googleSub` to match Google's terminology.
                googleSub: user.googleId,
            },
        }
    }
}

File: 3-oauth2-google-login/backend/1-java/src/main/java/com/example/oauthdemo/service/AuthService.java
package com.example.oauthdemo.service;

import com.example.oauthdemo.entity.User;
import com.example.oauthdemo.repository.UserRepository;
import com.example.oauthdemo.security.JwtService;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Business logic for the Google OAuth2 login flow.
 *
 * <p>Owns two responsibilities:
 * <ol>
 *   <li>Federated identity upsert — map a Google profile to a local {@link User} row.</li>
 *   <li>JWT issuance — sign an internal access token once the identity is confirmed.</li>
 * </ol>
 *
 * <p>Stateless service — no session or request-scoped state is held here.
 */
@Service
public class AuthService {
    /** Repository used to read and persist user rows. */
    private final UserRepository userRepository;
    /** Handles HMAC-SHA256 JWT signing. */
    private final JwtService jwtService;

    /**
     * Constructs the service with required dependencies injected by Spring.
     *
     * @param userRepository JPA repository for {@link User} persistence.
     * @param jwtService     helper that signs internal JWTs.
     */
    public AuthService(UserRepository userRepository, JwtService jwtService) {
        this.userRepository = userRepository;
        this.jwtService = jwtService;
    }

    /**
     * Upserts a local user row linked to the provided Google identity.
     *
     * <p>Lookup strategy (two-stage):
     * <ol>
     *   <li>Search by {@code googleId} — the stable provider identifier for returning users.</li>
     *   <li>If not found, search by {@code email} — to silently link an existing local account with Google.</li>
     *   <li>If neither exists, INSERT a new row with {@code password = null} (OAuth-only account).</li>
     * </ol>
     *
     * @param googleId  stable Google subject identifier ({@code sub} claim); used as primary upsert key.
     * @param email     verified email address; used as fallback upsert key and stored on the row.
     * @param firstName user's given name from the Google profile; may be {@code null}.
     * @param lastName  user's family name from the Google profile; may be {@code null}.
     * @param picture   URL to the user's Google profile photo; may be {@code null}.
     * @return a {@link Map} containing {@code "user"} ({@link User}) and {@code "isNewUser"} ({@link Boolean}).
     */
    public Map<String, Object> findOrCreateGoogleUser(String googleId, String email, String firstName, String lastName, String picture) {
        // Stage 1: try stable googleId first — fastest path for returning users.
        Optional<User> userOpt = userRepository.findByGoogleId(googleId);
        if (userOpt.isEmpty()) {
            // Stage 2: fall back to email so a pre-existing password account can gain Google login.
            userOpt = userRepository.findByEmail(email);
        }

        User user;
        boolean isNewUser;
        if (userOpt.isEmpty()) {
            // First time we see this identity — INSERT a new OAuth-only user row.
            // password is null because the credential lives at Google, not in our database.
            user = new User();
            user.setEmail(email);
            user.setGoogleId(googleId);
            user.setFirstName(firstName);
            user.setLastName(lastName);
            user.setPicture(picture);
            user.setPassword(null); // OAuth-only account: no local password
            user = userRepository.save(user);
            isNewUser = true;
        } else {
            // Returning user — re-link googleId (handles email-matched case where googleId was null).
            user = userOpt.get();
            user.setGoogleId(googleId); // always overwrite to guarantee the link is set
            // Only overwrite name/picture if Google provided a non-null value.
            if (firstName != null) user.setFirstName(firstName);
            if (lastName != null) user.setLastName(lastName);
            if (picture != null) user.setPicture(picture);
            user = userRepository.save(user);
            isNewUser = false;
        }

        // Wrap result in a map so the controller can extract user and isNewUser separately.
        Map<String, Object> result = new HashMap<>();
        result.put("user", user);
        result.put("isNewUser", isNewUser);
        return result;
    }

    /**
     * Signs an internal JWT and formats the final JSON response returned to the client.
     *
     * <p>The JWT carries only the {@code sub} (user id) claim — all other data is fetched
     * fresh from the database on each authenticated request, avoiding stale-claim issues.
     *
     * @param user      persisted user entity whose id becomes the JWT subject.
     * @param isNewUser {@code true} when the user row was just INSERTed; {@code false} on upsert.
     * @return map with keys {@code message}, {@code access_token}, {@code isNewUser}, and {@code user}.
     */
    public Map<String, Object> completeGoogleLogin(User user, boolean isNewUser) {
        // Sign JWT with only the user id — minimal payload reduces token size.
        String accessToken = jwtService.generateToken(user.getId());

        // Build the display name by joining non-empty first/last name parts.
        String name = Stream.of(user.getFirstName(), user.getLastName())
                .filter(s -> s != null && !s.isEmpty())
                .collect(Collectors.joining(" "));

        // Safe user summary — never expose password hash or internal db fields.
        Map<String, Object> userMap = new HashMap<>();
        userMap.put("id", user.getId());
        userMap.put("email", user.getEmail());
        userMap.put("name", name);
        userMap.put("googleSub", user.getGoogleId()); // alias matching Google's "sub" terminology

        // Assemble the full response matching the API contract shared across all four lang backends.
        Map<String, Object> response = new HashMap<>();
        response.put("message", "Đăng nhập Google thành công!");
        response.put("access_token", accessToken);
        response.put("isNewUser", isNewUser);
        response.put("user", userMap);

        return response;
    }
}
